README.md
# geekbot-api-py
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-3-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

A Geekbot (<https://geekbot.com/>) API client in python supporting async and sync operations.

Implements the Geekbot API per <https://geekbot.com/developers/>

<p align="center">
    <a href="https://github.com/andrewthetechie/geekbot-api-py" target="_blank">
        <img src="https://img.shields.io/github/last-commit/andrewthetechie/geekbot-api-py" alt="Latest Commit">
    </a>
        <img src="https://github.com/andrewthetechie/geekbot-api-py/actions/workflows/run_tests_with_tox.yaml/badge.svg?branch=main">
        <img src="https://codecov.io/gh/andrewthetechie/geekbot-api-py/branch/main/graph/badge.svg?token=7BK6JL2VL4"/>
    <br />
    <a href="https://pypi.org/project/geekbot-api" target="_blank">
        <img src="https://img.shields.io/pypi/v/geekbot-api" alt="Package version">
    </a>
    <img src="https://img.shields.io/pypi/pyversions/geekbot-api">
    <img src="https://img.shields.io/badge/license-MIT-green">
</p>

## Main Dependencies

- [Python +3.7](https://www.python.org)
- [pydantic](https://github.com/samuelcolvin/pydantic/)
- [httpx](https://www.python-httpx.org/)
- [requests](https://docs.python-requests.org/en/master/)

## Getting Started

### Installation

Install the package

    pip install geekbot-api

### Usage

Follow the directions on <https://geekbot.com/developers/> to get an API Token

#### Example

    from geekbot_api.config import GeekbotAPIConfig
    from geekbot_api.client import GeekbotAPIClient

    config = GeekbotAPIConfig(api_key="api_YOURKEYGOESHERE")

    client = GeekbotAPIClient(config=config)

    for standup in client.standups.list():
        standups.append(standup)

        print(standups)

#### Async Example

    import asyncio
    from geekbot_api.config import GeekbotAPIConfig
    from geekbot_api.client import GeekbotAPIClient

    config = GeekbotAPIConfig(api_key="api_YOURKEYGOESHERE")

    client = GeekbotAPIClient(config=config)

    async def print_standups():
        standups = list()
        async for standup in client.async_standups.list():
            standups.append(standup)

        print(standups)

    if __name__ == "__main__":
        loop = asyncio.get_event_loop()
        loop.run_until_complete(print_standups())

#### Reporting to a standup using schemas

    import asyncio
    from geekbot_api.config import GeekbotAPIConfig
    from geekbot_api.client import GeekbotAPIClient
    from geekbot_api.schemas import ReportIn

    config = GeekbotAPIConfig(api_key="api_YOURKEYGOESHERE")

    client = GeekbotAPIClient(config=config)
    standup_id = "12345" # Your standup id
    answers = dict() # keys are question ids, values are a dict responding to the question
    answers['123'] = {'text': "My first answer"}
    answers['234'] = {'text': "My second answer"}

    report_in = ReportIn(standup_id=standup_id, answers=answers) # pydantic will validate your inputs
    report = client.reports.create(report_in)
    print(report)

## Development

The [Makefile](./makefile) has useful targets to help setup your
development encironment. We suggest using pyenv to have access to
multiple python versions easily.

### Environment Setup

- Clone the repo and enter its root folder

  ```{.sourceCode .bash}
  git clone https://github.com/andrewthetechie/geekbot-api-py.git && cd geekbot-api-py
  ```

- Create a python 3.11 virtual environment and activate it. We suggest
  using [pyenv](https://github.com/pyenv/pyenv) to easily setup
  multiple python environments on multiple versions.

- Install the dependencies

  ```{.sourceCode .bash}
  make setup
  ```

### How to Run Tests

- Run the test command to run tests on only python 3.9

  ```{.sourceCode .bash}
  pytest
  ```

- Run Nox to run all python version tests

  ```{.sourceCode .bash}
  nox -s tests
  ```

### Test Requirements

Prs should always have tests to cover the change being made. Code
coverage goals for this project are 100% coverage.

### Code Linting

Code is linted with ruff

You can run the linting manually with make

```{.sourceCode .bash}
make lint
```

## CI

CI is run via Github Actions on all PRs and pushes to the main branch.

Releases are automatically released by Github Actions to Pypi.

## License

Licensed under the [MIT License](./LICENSE)

### Contributors

Thanks go to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tbody>
    <tr>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/andrewthetechie"><img src="https://avatars.githubusercontent.com/u/1377314?v=4?s=100" width="100px;" alt="Andrew"/><br /><sub><b>Andrew</b></sub></a><br /><a href="https://github.com/andrewthetechie/geekbot-api-py/commits?author=andrewthetechie" title="Code">💻</a> <a href="https://github.com/andrewthetechie/geekbot-api-py/commits?author=andrewthetechie" title="Documentation">📖</a> <a href="https://github.com/andrewthetechie/geekbot-api-py/commits?author=andrewthetechie" title="Tests">⚠️</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/CodingDavid8"><img src="https://avatars.githubusercontent.com/u/54584421?v=4?s=100" width="100px;" alt="David"/><br /><sub><b>David</b></sub></a><br /><a href="https://github.com/andrewthetechie/geekbot-api-py/commits?author=CodingDavid8" title="Code">💻</a> <a href="https://github.com/andrewthetechie/geekbot-api-py/issues?q=author%3ACodingDavid8" title="Bug reports">🐛</a></td>
      <td align="center" valign="top" width="14.28%"><a href="https://github.com/rmathew8-gh"><img src="https://avatars.githubusercontent.com/u/117225085?v=4?s=100" width="100px;" alt="rmathew8-gh"/><br /><sub><b>rmathew8-gh</b></sub></a><br /><a href="https://github.com/andrewthetechie/geekbot-api-py/commits?author=rmathew8-gh" title="Code">💻</a> <a href="https://github.com/andrewthetechie/geekbot-api-py/commits?author=rmathew8-gh" title="Tests">⚠️</a></td>
    </tr>
  </tbody>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
# Changelog

## [0.2.1](https://github.com/andrewthetechie/geekbot-api-py/compare/v0.2.0...v0.2.1) (2023-02-25)


### Bug Fixes

* update reports schema and add tests ([#21](https://github.com/andrewthetechie/geekbot-api-py/issues/21)) ([0e3b5ee](https://github.com/andrewthetechie/geekbot-api-py/commit/0e3b5eee18fa1c909277579279b5d627bdf7678d))


### Documentation

* add andrewthetechie as a contributor for code, doc, and test ([#24](https://github.com/andrewthetechie/geekbot-api-py/issues/24)) ([1989b9f](https://github.com/andrewthetechie/geekbot-api-py/commit/1989b9f244323b90e9d64226f274e3063076fd0e))
* add CodingDavid8 as a contributor for code, and bug ([#25](https://github.com/andrewthetechie/geekbot-api-py/issues/25)) ([b0c4a7e](https://github.com/andrewthetechie/geekbot-api-py/commit/b0c4a7edc12b0eb7f6faf1473ff4971b66930509))
* add rmathew8-gh as a contributor for code, and test ([#26](https://github.com/andrewthetechie/geekbot-api-py/issues/26)) ([1f3d0da](https://github.com/andrewthetechie/geekbot-api-py/commit/1f3d0da0dbdbd2a6305cc51de89160917a31a8d4))
* update readme with allcontributors block ([f86c3fd](https://github.com/andrewthetechie/geekbot-api-py/commit/f86c3fdc4f6a91259e2b368c86b7174926162fe8))

## [0.2.0](https://github.com/andrewthetechie/geekbot-api-py/compare/v0.1.0...v0.2.0) (2023-02-21)


### Features

* fix version and package name ([5a87dcd](https://github.com/andrewthetechie/geekbot-api-py/commit/5a87dcd6a98e3037d387df8c7107d1bee504e311))
* move to poetry ([#12](https://github.com/andrewthetechie/geekbot-api-py/issues/12)) ([525154d](https://github.com/andrewthetechie/geekbot-api-py/commit/525154d0cc94cdc9569bac6b59d35c3e19d112f5))


### Bug Fixes

* 3: StandupBase requiring personalized ([#4](https://github.com/andrewthetechie/geekbot-api-py/issues/4)) ([9bdcaa9](https://github.com/andrewthetechie/geekbot-api-py/commit/9bdcaa900cf154f9951ca9f055ec924db7d623eb))
* bump requirements ([#6](https://github.com/andrewthetechie/geekbot-api-py/issues/6)) ([966bfa0](https://github.com/andrewthetechie/geekbot-api-py/commit/966bfa0f610fbe6219fe5e209d35e40f16115861))
* fix create report ([44fbc72](https://github.com/andrewthetechie/geekbot-api-py/commit/44fbc725bc0379fd860092a372c27b1abd04ba18))
* fix schemas ([382df3f](https://github.com/andrewthetechie/geekbot-api-py/commit/382df3fe9699ee2feefbeb28c82824378a27959c))
* include clients in packaging ([0497ca3](https://github.com/andrewthetechie/geekbot-api-py/commit/0497ca3d5f7ae2c361d0a91f49fc0baa8689b7f6))


noxfile.py
"""Nox sessions."""
import os
import shlex
import sys
from pathlib import Path
from textwrap import dedent

import nox
import toml

try:
    from nox_poetry import Session
    from nox_poetry import session
except ImportError:
    message = f"""\
    Nox failed to import the 'nox-poetry' package.

    Please install it using the following command:

    {sys.executable} -m pip install nox-poetry"""
    raise SystemExit(dedent(message)) from None


package = "geekbot_api"
python_versions = [
    "3.10",
    "3.9",
    "3.8",
    "3.7",
    "3.11",
]
nox.needs_version = ">= 2021.6.6"
nox.options.sessions = (
    "pre-commit",
    "bandit",
    "safety",
    # "mypy", Turn off mypy for now, lots of errors
    "tests",
)
mypy_type_packages = ()
pyproject = toml.load("pyproject.toml")
test_requirements = pyproject["tool"]["poetry"]["group"]["dev"]["dependencies"].keys()


def activate_virtualenv_in_precommit_hooks(session: Session) -> None:
    """Activate virtualenv in hooks installed by pre-commit.

    This function patches git hooks installed by pre-commit to activate the
    session's virtual environment. This allows pre-commit to locate hooks in
    that environment when invoked from git.

    Args:
        session: The Session object.
    """
    assert session.bin is not None  # noqa: S101

    # Only patch hooks containing a reference to this session's bindir. Support
    # quoting rules for Python and bash, but strip the outermost quotes so we
    # can detect paths within the bindir, like <bindir>/python.
    bindirs = [
        bindir[1:-1] if bindir[0] in "'\"" else bindir for bindir in (repr(session.bin), shlex.quote(session.bin))
    ]

    virtualenv = session.env.get("VIRTUAL_ENV")
    if virtualenv is None:
        return

    headers = {
        # pre-commit < 2.16.0
        "python": f"""\
            import os
            os.environ["VIRTUAL_ENV"] = {virtualenv!r}
            os.environ["PATH"] = os.pathsep.join((
                {session.bin!r},
                os.environ.get("PATH", ""),
            ))
            """,
        # pre-commit >= 2.16.0
        "bash": f"""\
            VIRTUAL_ENV={shlex.quote(virtualenv)}
            PATH={shlex.quote(session.bin)}"{os.pathsep}$PATH"
            """,
    }

    hookdir = Path(".git") / "hooks"
    if not hookdir.is_dir():
        return

    for hook in hookdir.iterdir():
        if hook.name.endswith(".sample") or not hook.is_file():
            continue

        if not hook.read_bytes().startswith(b"#!"):
            continue

        text = hook.read_text()

        if not any(Path("A") == Path("a") and bindir.lower() in text.lower() or bindir in text for bindir in bindirs):
            continue

        lines = text.splitlines()

        for executable, header in headers.items():
            if executable in lines[0].lower():
                lines.insert(1, dedent(header))
                hook.write_text("\n".join(lines))
                break


@session(name="pre-commit", python=python_versions[0])
def precommit(session: Session) -> None:
    """Lint using pre-commit."""
    args = session.posargs or ["run", "--all-files", "--show-diff-on-failure"]
    session.install(*test_requirements)
    session.install(".")
    session.run("pre-commit", *args)
    if args and args[0] == "install":
        activate_virtualenv_in_precommit_hooks(session)


@session(python=python_versions[0])
def safety(session: Session) -> None:
    """Scan dependencies for insecure packages."""
    requirements = session.poetry.export_requirements()
    session.install("safety")
    # ignore https://github.com/pytest-dev/py/issues/287
    # its an irresposnbily filed CVE causing nose
    session.run("safety", "check", "--full-report", f"--file={requirements}", "--ignore=51457")


@session(python=python_versions)
def mypy(session: Session) -> None:
    """Type-check using mypy."""
    args = session.posargs or ["src/geekbot_api"]
    session.install(".")
    session.install("mypy", "pytest")
    if len(mypy_type_packages) > 0:
        session.install(*mypy_type_packages)
    session.run("mypy", *args)


@session(python=python_versions[0])
def bandit(session: Session) -> None:
    """Run bandit security tests"""
    session.install("bandit")
    args = session.posargs or ["-r", "./geekbot_api"]
    session.run("bandit", *args)


@session(python=python_versions)
def tests(session: Session) -> None:
    """Run the test suite."""
    session.install(".")
    session.install(*test_requirements)
    session.run("poetry", "run", "pytest", *session.posargs)
